lib/teleterm/clusters/cluster_auth.go

```diff
@@ -96,8 +96,29 @@ func (c *Cluster) LocalLogin(ctx context.Context, user, password, otpToken strin
 		if err != nil {
 			return trace.Wrap(err)
 		}
-	case constants.SecondFactorU2F, constants.SecondFactorWebauthn, constants.SecondFactorOn, constants.SecondFactorOptional:
+	case constants.SecondFactorU2F, constants.SecondFactorWebauthn:
 		err := c.localMFALogin(ctx, user, password)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	case constants.SecondFactorOn, constants.SecondFactorOptional:
+		var err error
+
+		// tsh always uses client.SSHAgentMFALogin for any `second_factor` option other than `off` and
+		// `otp`. If it's set to `on` or `optional` and it turns out the user wants to use an OTP, it
+		// bails out to stdin to ask them for it.
+		//
+		// Connect cannot do that, but it still wants to use the auth code from lib/client that it
+		// shares with tsh. So to temporarily work around this problem, we check if the OTP token was
```

Comment:
Do we have a plan to properly fix it (vs "temporarily")?

Reply:
Thanks for the tips! At the moment we do have a lot of code copied from tsh. But knowing what to do with the auth methods will help us in deciding which parts of the code put inside `lib` and share and which parts should actually stay separate.

Reply:
@codingllama I looked very briefly at doing webauthn through Chromium in Electron and it might be tricky at the moment because in general webauthn in browsers requires websites to use HTTPS (see https://github.com/electron/electron/issues/24573).

Also, my first reaction to the suggestion of not using tsh-specific methods was "Ahh, but tsh already does a lot of work with the keys it gets in response from that end point, it'd be a shame if we were not able to use that code 😢". What I didn't fully realize at the time is that actually going through the MFA challenge and then processing the response are two distinct steps and [we actually just copy-pasted](https://github.com/gravitational/teleport/blob/2f00cb4670d0fca23419d8229843134d7d4885a3/lib/teleterm/clusters/cluster_auth.go#L232) the response-processing code [from tsh](https://github.com/gravitational/teleport/blob/2f00cb4670d0fca23419d8229843134d7d4885a3/lib/client/api.go#L2574-L2596).

So down the road it should be possible to directly talk to the proxy API and use native browser APIs for webauthn. It's all down to what Electron will allow us to do, but I didn't check that and don't plan to for now.

I'm looking at this atm because we're talking about adding per-session MFA support to Connect which requires the user interacting with a hardware key. And I remembered that the current solution is a bit hacky.